src/core/connection.c

```diff
@@ -1892,6 +1892,223 @@ QuicConnRestart(
+    QUIC_TRANSPORT_PARAMETERS LocalTPCopy;
+    QuicCopyMemory(&LocalTPCopy, &Connection->LocalTP, sizeof(LocalTPCopy));
+    LocalTPCopy.Flags = LocalTPCopy.Flags & (
+        QUIC_TP_FLAG_ACTIVE_CONNECTION_ID_LIMIT |
+        QUIC_TP_FLAG_INITIAL_MAX_DATA |
+        QUIC_TP_FLAG_INITIAL_MAX_STRM_DATA_BIDI_LOCAL |
+        QUIC_TP_FLAG_INITIAL_MAX_STRM_DATA_BIDI_REMOTE |
+        QUIC_TP_FLAG_INITIAL_MAX_STRM_DATA_UNI |
+        QUIC_TP_FLAG_INITIAL_MAX_STRMS_BIDI |
+        QUIC_TP_FLAG_INITIAL_MAX_STRMS_UNI);
+
+    const uint8_t* EncodedLocalTP = QuicCryptoTlsEncodeTransportParameters(
+        Connection,
+        &LocalTPCopy,
+        &EncodedTransportParametersLength);
+
+    if (EncodedLocalTP == NULL) {
+        Status = QUIC_STATUS_OUT_OF_MEMORY;
+        goto Error;
+    }
+    Ticket.TransportParamsLength = EncodedTransportParametersLength;
+
+    uint32_t TotalTicketLength = (uint32_t) (QuicVarIntSize(Ticket.TicketVersion) +
+        QuicVarIntSize(Ticket.AlpnLength) +
+        QuicVarIntSize(Ticket.TransportParamsLength) +
+        QuicVarIntSize(Ticket.AppTicketLength) +
+        sizeof(Ticket.QuicVersion) +
+        Ticket.AlpnLength +
+        Ticket.TransportParamsLength +
+        Ticket.AppTicketLength);
+
+    TicketBuffer = QUIC_ALLOC_NONPAGED(TotalTicketLength);
+    if (TicketBuffer == NULL) {
+        QuicTraceEvent(
+            AllocFailure,
+            "Allocation of '%s' failed. (%llu bytes)",
+            "Server resumption ticket",
+            TotalTicketLength);
+        Status = QUIC_STATUS_OUT_OF_MEMORY;
+        goto Error;
+    }
+
+    /*
+        Encoded ticket format is as follows:
+        Ticket Version (QUIC_VAR_INT) [1..4]
+        Quic Version [4]
+        Negotiated ALPN length (QUIC_VAR_INT) [1..4]
+        Negotiated ALPN [...]
+        Transport Parameters length (QUIC_VAR_INT) [1..4]
+        Transport Parameters [...]
+        App Ticket length (QUIC_VAR_INT) [1..4]
+        App Ticket (omitted if length is zero) [...]
+    */
+    uint8_t* TicketCursor = QuicVarIntEncode(Ticket.TicketVersion, TicketBuffer);
+    *(uint32_t*)TicketCursor = QuicByteSwapUint32(Ticket.QuicVersion);
+    TicketCursor += sizeof(Ticket.QuicVersion);
+    TicketCursor = QuicVarIntEncode(Ticket.AlpnLength, TicketCursor);
+    QuicCopyMemory(TicketCursor, Connection->Crypto.TlsState.NegotiatedAlpn + 1, (size_t) Ticket.AlpnLength);
+    TicketCursor += Ticket.AlpnLength;
+    TicketCursor = QuicVarIntEncode(Ticket.TransportParamsLength, TicketCursor);
+    QuicCopyMemory(TicketCursor, EncodedLocalTP, (size_t) Ticket.TransportParamsLength);
+    TicketCursor += EncodedTransportParametersLength;
+    TicketCursor = QuicVarIntEncode(Ticket.AppTicketLength, TicketCursor);
+    if (Ticket.AppTicketLength > 0) {
+        QuicCopyMemory(TicketCursor, AppResumptionData, (size_t) Ticket.AppTicketLength);
+    }
+
+    Status = QuicTlsSendTicket(
+        Connection->Crypto.TLS,
+        TotalTicketLength,
+        TicketBuffer);
```

Comment:
I'm a bit confused as to how this then causes the CRYPTO frame to be sent out. What goes and drains the TLS data after this?

Reply:
The TLS layer issues a `ProcessCompleteCallback` callback to indicate that there is data ready. The existing infrastructure for reading data from TLS and packaging and sending CRYPTO frames handles it from there.